.ci/magician/cmd/generate_downstream.go

```diff
@@ -0,0 +1,338 @@
+
+	It expects the following arguments:
+	1. Command, either head, base, or downstream
+	2. Name of the downstream repo, either terraform, terraform-google-conversion, or tf-oics
+	3. Version of the downstream
+	4. Commit SHA of the squashed merge commit
+
+	The following environment variables should be set:
+` + listGDEnvironmentVariables(),
+	Run: func(cmd *cobra.Command, args []string) {
+		env := make(map[string]string, len(gdEnvironmentVariables))
+		for _, ev := range gdEnvironmentVariables {
+			val, ok := os.LookupEnv(ev)
+			if !ok {
+				fmt.Printf("Did not provide %s environment variable\n", ev)
+				os.Exit(1)
+			}
+			env[ev] = val
+		}
+
+		gh := github.NewClient()
+		rnr, err := exec.NewRunner()
+		if err != nil {
+			fmt.Println("Error creating a runner: ", err)
+			os.Exit(1)
+		}
+		ctlr := source.NewController(env["GOPATH"], "modular-magician", env["GITHUB_TOKEN"], rnr)
+
+		if len(args) != 4 {
+			fmt.Printf("Wrong number of arguments %d, expected 4\n", len(args))
+			os.Exit(1)
+		}
+
+		execGenerateDownstream(env["BASE_BRANCH"], args[0], args[1], args[2], args[3], gh, rnr, ctlr)
+	},
+}
+
+func listGDEnvironmentVariables() string {
+	var result string
+	for i, ev := range gdEnvironmentVariables {
+		result += fmt.Sprintf("\t%2d. %s\n", i+1, ev)
+	}
+	return result
+}
+
+func execGenerateDownstream(baseBranch, command, repo, version, ref string, gh GithubClient, rnr *exec.Runner, ctlr *source.Controller) {
+	if baseBranch == "" {
+		baseBranch = "main"
+	}
+	upstreamRepo, scratchRepo, err := cloneRepo(baseBranch, repo, version, ctlr)
+	if err != nil {
+		fmt.Println("Error cloning repo: ", err)
+		os.Exit(1)
+	}
+
+	mmLocalPath := filepath.Join(rnr.GetCWD(), "..", "..")
+	mmCopyPath := filepath.Join(mmLocalPath, "..", fmt.Sprintf("mm-%s-%s-%s", repo, version, command))
+	if _, err := rnr.Run("cp", []string{"-rp", mmLocalPath, mmCopyPath}, nil); err != nil {
+		fmt.Println("Error copying magic modules: ", err)
+		os.Exit(1)
+	}
+
+	if err := rnr.PushDir(mmCopyPath); err != nil {
+		fmt.Println("Error changing directory to copied magic modules: ", err)
+		os.Exit(1)
+	}
+
+	if err := setConfig(rnr); err != nil {
+		fmt.Println("Error setting config: ", err)
+		os.Exit(1)
+	}
+
+	mmRepo := &source.Repo{
+		Name: "magic-modules",
+		Path: mmCopyPath,
+	}
+	commitMessage, err := branchAndCommitMessage(mmRepo, scratchRepo, command, ref, rnr, ctlr)
+	if err != nil {
+		fmt.Println("Error getting branch and commit message: ", err)
+		os.Exit(1)
+	}
+
+	if err := runMake(upstreamRepo, command, rnr); err != nil {
+		fmt.Println("Error running make: ", err)
+		os.Exit(1)
+	}
+
+	if err := pushCommit(upstreamRepo, scratchRepo, baseBranch, command, commitMessage, ref, gh, rnr, ctlr); err != nil {
+		fmt.Println("Error pushing commit: ", err)
+		os.Exit(1)
+	}
+}
+
+func cloneRepo(baseBranch, repo, version string, ctlr *source.Controller) (*source.Repo, *source.Repo, error) {
+	upstreamRepo := &source.Repo{
```

Comment:
```suggestion
	downstreamRepo := &source.Repo{
```

downstream is indicative of the relationship between these two repos.

Reply:
I understand the original script used this terminology but in retrospect it seems incorrect.

https://github.com/GoogleCloudPlatform/magic-modules/blob/5f5f9b3fee520b56f9dae6ec164d855d9074df87/.ci/scripts/build-environment/downstream-builder/generate_downstream.sh#L8

Reply:
Done